src/Interpreters/ProcessList.cpp

```diff
@@ -137,31 +138,28 @@ ProcessList::EntryPtr ProcessList::insert(
         IAST::QueryKind query_kind = ast ? ast->getQueryKind() : IAST::QueryKind::Select;
 
         const auto queue_max_wait_ms = settings[Setting::queue_max_wait_ms].totalMilliseconds();
-        UInt64 waiting_queries = waiting_queries_amount.load();
-        if (!is_unlimited_query && max_size && processes.size() >= max_size + waiting_queries)
+        if (!is_unlimited_query && max_size && countNonInternalProcesses() >= max_size)
```

Comment:
I wonder what the performance impact of iterating over the list for every query will be with thousands of concurrent queries. Same for the wait below or the check for `max_concurrent_queries_for_all_users`.

Did you consider having separate lists (or a better data structure) to avoid it?

Reply:
> I wonder what the performance impact of iterating over the list for every query will be with thousands of concurrent queries.

It should be in the tens of microseconds range, but I still added an extra counter just to be safe.